geode-core/src/integrationTest/java/org/apache/geode/internal/cache/TXDetectReadConflictJUnitTest.java

```diff
@@ -88,204 +92,176 @@ public void tearDown() throws Exception {
     mgr.begin();
-    assertEquals("value", region.get("key"));
-    assertEquals("value1", region.get("key1"));
+    assertEquals(value, region.get(key));
+    assertEquals(value1, region.get(key1));
     mgr.commit();
   }
 
   @Test
   public void testReadConflictsPR() throws Exception {
     cache.close();
     createCachePR();
-    regionpr.put("key", "value");
-    regionpr.put("key1", "value1");
+    regionPR.put(key, value);
+    regionPR.put(key1, value1);
     TXManagerImpl mgr = (TXManagerImpl) cache.getCacheTransactionManager();
     mgr.begin();
-    assertEquals("value", regionpr.get("key"));
-    assertEquals("value1", regionpr.get("key1"));
+    assertEquals(value, regionPR.get(key));
+    assertEquals(value1, regionPR.get(key1));
     mgr.commit();
   }
 
 
   /**
-   * Test that two transactions with only read operations don't produce CommitConflictException
-   * This test fill some initial value on startup and after that create different threads with
-   * region#get operations
-   * Sync two different threads commit time through lock
+   * Test that two transactions with read and put operations produce CommitConflictException
    */
-  @Test(/* no exception expected */)
-  public void testDetectReadConflict()
-      throws CacheException, ExecutionException, InterruptedException {
+
+  @Test
+  public void readConflictsTransactionCanBlockWriteTransaction() {
     cache.close();
     createCache();
-    TXManagerImpl txMgrImpl = (TXManagerImpl) cache.getCacheTransactionManager();
-    // fill some initial key-value on start
-    txMgrImpl.begin();
-    this.region.put("key1", "value1"); // non-tx
-    txMgrImpl.commit();
-    ExecutorService executor = Executors.newFixedThreadPool(2);
-    for (int i = 0; i < 20; i++) {
-      final StampedLock lock = new StampedLock();
-      lock.asWriteLock().lock();
-      //
-      Future<?> future1 = executor.submit(() -> {
-        CacheTransactionManager txMgr2 = cache.getCacheTransactionManager();
-        txMgr2.begin();
-        region.get("key1"); // non-tx
-        lock.asReadLock().lock();
-        txMgr2.commit();
-      });
-      Future<?> future2 = executor.submit(() -> {
-        CacheTransactionManager txMgr3 = cache.getCacheTransactionManager();
-        txMgr3.begin();
-        region.get("key1"); // non-tx
-        lock.asReadLock().lock();
-        txMgr3.commit();
-      });
-      pause(100);
-      lock.asWriteLock().unlock();
-      future1.get();
-      future2.get();
+
+    region.put(key, value);
+    region.put(key1, value1);
+    TXManagerImpl txManager = (TXManagerImpl) cache.getCacheTransactionManager();
+    txManager.begin();
+    assertThat(region.get(key)).isSameAs(value);
+    region.put(key1, newValue1);
+    TXState txState =
+        (TXState) ((TXStateProxyImpl) TXManagerImpl.getCurrentTXState()).getRealDeal(null, null);
+    txState.setAfterReservation(() -> readTransactionAfterReservation());
+    Runnable task = () -> doPutOnReadKeyTransaction();
```

Comment:
Please use ExecutorServiceRule.ThrowingRunnable.

You can see the original review comments on PR #3742.